swagger_server/itm/itm_scenario.py

```diff
@@ -1,73 +1,240 @@
+        self.session.history.add_history(
+            "Change scene",
+            {"session_id": self.session.session_id,
+            "scenario_id": self.id,
+            "scene_index": self.isd.current_scene_index},
+            target_state.to_dict() if target_state else None
+        )
+
+        '''
+        Merge state from new scene into session.state.  Approach:
+        0. Abort if no state to merge
+        1. Always replace entire `characters` structure.
+        2. For `supplies`, add or update any specified supplies.
+        3. For everything else, replace any specified (non-None) values
+           3a. Lists are copied whole (e.g., `character_importance`, `aid_delay`, `threats`).
+        4. Clear hidden data (e.g., character vitals)
+        '''
+        # Rule 0: Abort if no state to merge
+        if not target_state:
+            current_state.characters = []
+            return
+
+        # Rule 1: Always replace entire `characters` structure.
+        current_state.characters = target_state.characters
+
+        # Rule 2: For `supplies`, add or update any specified supplies.
+        if target_state.supplies:
+            target_types = []
+            current_types = []
+            for target_supply in target_state.supplies:
+                target_types.append(target_supply.type)
+                for current_supply in current_state.supplies:
+                    current_types.append(current_supply.type)
+                    if target_supply.type == current_supply.type:
+                        current_supply.quantity = target_supply.quantity
+                        current_supply.reusable = target_supply.reusable
+            new_types = [new_type for new_type in target_types if new_type not in current_types]
+            new_supplies = [new_supply for new_supply in target_state.supplies if new_supply.type in new_types]
+            current_state.supplies.extend(new_supplies)
+
+        # Rule 3: For everything else, replace any specified (non-None) values.
+        # Lists are copied whole (e.g., `character_importance`, `aid_delay`, `threats`).
+
+        # Top-level unstructured.
+        if target_state.unstructured:
+            current_state.unstructured = target_state.unstructured
+
+        # Threat state
+        if target_state.threat_state:
+            target = target_state.threat_state
+            current = current_state.threat_state
+            if not current:
+                current = target
+            else:
+                current.unstructured=target.unstructured if target.unstructured else current.unstructured
+                current.threats = target.threats if target.threats else current.threats
+            current_state.threat_state = current
+
+        # Mission
+        if target_state.mission:
+            target = target_state.mission
+            current = current_state.mission
+            if not current:
+                current = target
+            else:
```

Comment:
I'm not sure if this list of attributes will continue to grow or be modified in the future but here is a way to maybe make this section a little more maintainable
```
attributes = [
        "unstructured", "mission_type", "character_importance", 
        "civilian_presence", "communication_capability", "medical_policies", 
        "political_climate", "roe"
    ]
   
    for attr in attributes:
        setattr(current, attr, getattr(target, attr) or getattr(current, attr))
    
current_state.mission = current
```
 This way you can just maintain a dictionary of all of the attributes to loop over instead of having to manually maintain this section.

Reply:
The same idea (if feasible here) applies to sim env and decision env below

Reply:
This looks really promising!  I knew there would be a slick Python-y way to do this, but I just needed to get something working out there.  :)  To add to your solution, I think we can use they keys from `Mission.attribute_map` (for example) that won't need any maintenance.  I'll try it out and test both ways and ensure the same results.

Reply:
As discussed in Teams, your solution work, as did my small addition mentioned above (except that I had to use an instance of the model object since `attribute_map` isn't static.

And once the changes were made, it became clear we could refactor it further.  Thank you!